persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java

```diff
@@ -594,161 +549,92 @@ ColumnFamilyDescriptor newDescriptor(byte[] name) {
-                      }
-                      writeBatch(batch);
-
-                      // Add metadata only after batch has been written
-                      for (MarshallableEntry entry : entries) {
-                          if (entry.expiryTime() > -1)
-                              addNewExpiry(entry);
-                      }
-                  })
-                  .doOnError(e -> {
-                      throw new PersistenceException(e);
-                  })
-                  .ignoreElements()
-                  .toCompletionStage(null);
-        }
-
-        void deleteBatch(Iterable<Object> keys) {
+        CompletionStage<Boolean> delete(int segment, Object key) {
             try {
-                int batchSize = 0;
-                WriteBatch batch = new WriteBatch();
-                for (Object key : keys) {
-                    batch.remove(getHandle(calculateSegment(key)), marshall(key));
-                    batchSize++;
-
-                    if (batchSize == configuration.maxBatchSize()) {
-                        batchSize = 0;
-                        writeBatch(batch);
-                        batch = new WriteBatch();
+                byte[] keyBytes = marshall(key);
+                ColumnFamilyHandle handle = getHandle(segment);
+                return blockingManager.supplyBlocking(() -> {
+                    try {
+                        if (db.get(handle, keyBytes) == null) {
+                            return Boolean.FALSE;
+                        }
+                        db.delete(handle, keyBytes);
+                        return Boolean.TRUE;
+                    } catch (RocksDBException e) {
+                        throw new PersistenceException(e);
                     }
-                }
-
-                if (batchSize != 0)
-                    writeBatch(batch);
+                }, "rocksdb-delete");
             } catch (Exception e) {
                 throw new PersistenceException(e);
             }
         }
 
-        abstract void clear(IntSet segments);
-
-        abstract Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter);
+        abstract CompletionStage<Void> clear();
 
         abstract Publisher<MarshallableEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter,
-                                                                   boolean fetchValue, boolean fetchMetadata);
-
-        int size(IntSet segments) {
-            CompletionStage<Long> stage = Flowable.fromPublisher(publishKeys(segments, null))
-                  .count().toCompletionStage();
+                boolean fetchValue);
 
-            long count = CompletionStages.join(stage);
-            if (count > Integer.MAX_VALUE) {
-                return Integer.MAX_VALUE;
-            }
-            return (int) count;
+        CompletionStage<Long> size(IntSet segments) {
+            return Flowable.fromPublisher(publishKeys(segments, null))
```

Comment:
For a segmented store there is no need to unmarshall all of the keys during the size calculation. Instead we can just return a  "empty" MarshalledEntry instance from `RocksEntryIterator` in order for `count()` to still be utilised.

Reply:
Unfortunately, expiration rears its ugly head again (we need to get at the metadata for expiration) - and at that point we require unmarshalling the value - so I don't think it is worth wasting code to not unmarshall the key.

This is one reason I was very glad to add in the `approximateSize` method (which won't work for non shared non segmented stores - shared or segmented should always work).